### PR TITLE
Implemented interface for esper query creation

### PR DESCRIPTION
**Pull request by friedow**

### CODE REVIEW

File: src/ProductView/ProductView.js [hunk 1]
```diff
@@ -105,6 +105,7 @@ class ProductView extends Component {
                         onChangeFilterInput={this.onChangeFilterInput}
                         filter={this.state.filter}/>
                     <TabBar
+                        dataSender={this.props.dataSender}
```
Comment (julianweise) [inline src/ProductView/ProductView.js]: Better naming suggestion: data reception

File: src/ProductView/TabBar/QueryInterface/Modal/Modal.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import React, { Component } from 'react';
```
Comment (julianweise) [inline src/ProductView/TabBar/QueryInterface/Modal/Modal.js]: Error handling + displaying is missing.

File: src/ProductView/TabBar/QueryInterface/Modal/Modal.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
```
Comment (julianweise) [inline src/ProductView/TabBar/QueryInterface/Modal/Modal.js]: Success notification is missing

File: src/ProductView/TabBar/QueryInterface/Modal/Modal.js (hunk 2 above)
Comment (julianweise) [reply src/ProductView/TabBar/QueryInterface/Modal/Modal.js]: clientDataContainer.clientErrorCallback is not a function

File: src/ProductView/TabBar/QueryInterface/QueryInterface.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        this.nextAttributeId = 3;
+        this.handleChangeEventTypeName = this.handleChangeEventTypeName.bind(this);
+        this.handleChangeAttributeName = this.handleChangeAttributeName.bind(this);
+        this.handleChangeAttributeType = this.handleChangeAttributeType.bind(this);
+        this.handleChangeQuery = this.handleChangeQuery.bind(this);
+        this.handleSaveQuery = this.handleSaveQuery.bind(this);
+    }
+
+    componentDidMount() {
+        this.setState({ eventQuery: this.props.defaultQuery });
```
Comment (julianweise) [inline src/ProductView/TabBar/QueryInterface/QueryInterface.js]: Please do not use props for defining defaultQueries.
Transfer it to a static method and after merging it to developer transfer it to config.

File: src/ProductView/TabBar/QueryInterface/QueryInterface.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            if (attributes[i].name) {
+                eventTypeAttributes[attributes[i].name] = attributes[i].type;
+            }
+        }
+        const eventType = {
+            'name': this.state.eventTypeName,
+            'timestamp': 'timestamp',
+            'attributes': eventTypeAttributes
+        };
+        return eventType;
```
Comment (julianweise) [inline src/ProductView/TabBar/QueryInterface/QueryInterface.js]: Please directly return builded object from eventType

File: src/ProductView/TabBar/QueryInterface/QueryInterface.js [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            "readonly": false
+        }]);
+        this.nextAttributeId = this.nextAttributeId + 1;
+        return attributes;
+    }
+
+    getEventType() {
+        const eventTypeAttributes = {};
+        const attributes = this.state.eventTypeAttributes;
+        for (let i = 0; i < attributes.length; i++) {
```
Comment (julianweise) [inline src/ProductView/TabBar/QueryInterface/QueryInterface.js]: Please comment on this construct.
I think it is necessary here, due to the special reuqirements of the backend (and JSON parsing) but it is 
poorly understandable.

File: src/ProductView/TabBar/QueryInterface/QueryInterface.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            'id': this.nextAttributeId,
+            "name": "",
+            "type": "",
+            "readonly": false
+        }]);
+        this.nextAttributeId = this.nextAttributeId + 1;
+        return attributes;
+    }
+
+    getEventType() {
```
Comment (julianweise) [inline src/ProductView/TabBar/QueryInterface/QueryInterface.js]: Give the function a more descriptivename. e.g.: buildBackendDataContainer or buildJSONSerializableDataContainer, ...

File: src/ProductView/TabBar/QueryInterface/QueryInterface.js [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        this.setState({ eventTypeAttributes: attributes });
+    }
+
+    handleChangeQuery(event) {
+        this.setState({ eventQuery: event.target.value });
+    }
+
+    handleSaveQuery() {
+        const eventType = this.getEventType();
+        this.props.dataSender.createEventtype(this.state.eventQuery, eventType, this.handleEventTypeData, this.handleError);
```
Comment (julianweise) [inline src/ProductView/TabBar/QueryInterface/QueryInterface.js]: line too long

File: src/ProductView/TabBar/QueryInterface/QueryInterface.js [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    render() {
+        return (
+            <div className="query-interface">
+                <Modal onSave={this.handleSaveQuery}>
+                    <QueryInterfaceHeader eventTypeName={this.state.eventTypeName}
```
Comment (julianweise) [inline src/ProductView/TabBar/QueryInterface/QueryInterface.js]: Maybe rename header to attributeList or something elese. Header indicates an protocol header or an UI header ...

File: src/ProductView/TabBar/QueryInterface/QueryInterfaceHeader/QueryInterfaceHeaderAttribute/QueryInterfaceHeaderAttribute.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+                           value={this.props.eventTypeAttribute.name}
+                           onChange={this.onChangeAttributeName}
+                           readOnly={this.props.eventTypeAttribute.readonly}/>
+                </div>
+                <div className="col-6">
+                    <select className="form-control" id="exampleSelect1"
+                            value={this.props.eventTypeAttribute.type}
+                            onChange={this.onChangeAttributeType}
+                            readOnly={this.props.eventTypeAttribute.readonly}>
+                        <option>STRING</option>
```
Comment (julianweise) [inline src/ProductView/TabBar/QueryInterface/QueryInterfaceHeader/QueryInterfaceHeaderAttribute/QueryInterfaceHeaderAttribute.js]: We should do this more generic.
e.g.: receiving information about possible data types from backend or transfering it to the central config.

File: src/ProductView/TabBar/TabBar.js [hunk 11]
```diff
@@ -12,6 +13,11 @@ class TabBar extends Component {
                              loadEventsFor={this.props.loadEventsFor}
                              product={this.props.product}/>
                     )}
+                    <li>
+                        <QueryInterface
+                            dataSender={this.props.dataSender}
+                            defaultQuery="INSERT INTO TestErrorEvents SELECT timestamp, productId, productFamilyId FROM FeedbackData"/>
```
Comment (julianweise) [inline src/ProductView/TabBar/TabBar.js]: See comment above about default query / values in source code.

File: src/RemoteHandler/DataTransmitter.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        const data = {
+            "eventQuery":   eventQuery,
+            "eventType":    eventType
+        };
+        const callbackContainer = {
+            "dataMappingFunction":    () => {/* Not necessary because no data mapping is needed. */},
+            "clientSuccessCallback":  successCallback,
+            "clientErrorCallback":    errorCallback
+        };
+        this.client.sendRequest(this.receiveResults, this.receiveError, callbackContainer, JSON.stringify(data));
```
Comment (julianweise) [inline src/RemoteHandler/DataTransmitter.js]: Line is too long

Comment (julianweise) [inline src/RemoteHandler/DataTransmitter.js]: I would suggest to create a new function for JSON parsing to catch parsing errors and displaying them sufficiently.

File: src/RemoteHandler/RESTInterface.js [hunk 13]
```diff
@@ -7,11 +7,11 @@ class RESTInterface {
         this.client.open(requestMethod, api, async);
     }
 
-    sendRequest(successCallback, errorCallback, clientDataContainer) {
+    sendRequest(successCallback, errorCallback, clientDataContainer, data = null) {
         this.client.onload = this.onSuccess.bind(this, successCallback, errorCallback, clientDataContainer);
         this.client.onerror = this.onError.bind(this, errorCallback, clientDataContainer);
         this.client.setRequestHeader("Content-type", "application/json");
-        this.client.send();
+        this.client.send(data);
```
Comment (julianweise) [inline src/RemoteHandler/RESTInterface.js]: Please check sending of null concerning any possible error case.

File: src/index.js [hunk 14]
```diff
@@ -14,15 +15,19 @@ import './index.scss';
 const API_SERVER_URL = "localhost";
 const API_SERVER_PORT = 8989;
 
-const dataSource = new ProductFetcher(API_SERVER_URL, API_SERVER_PORT);
+const dataSource = new DataReceiver(API_SERVER_URL, API_SERVER_PORT);
+const dataSender = new DataTransmitter(API_SERVER_URL, API_SERVER_PORT);
```
Comment (julianweise) [inline src/index.js]: See my comment above concerning naming.

**Review by julianweise — CHANGES_REQUESTED** [review]

File: src/ProductView/ProductView.js (hunk 1 above)
Comment (friedow) [reply src/ProductView/ProductView.js]: I think this name doesn't fit to well.

Comment (julianweise) [reply src/ProductView/ProductView.js]: yes, fair

**Comment by julianweise** [comment]
What´s about this pull request? It belongs to our last sprint and is still not solved.

File: src/RemoteHandler/DataTransmitter.js (hunk 12 above)
Comment (friedow) [inline src/RemoteHandler/DataTransmitter.js]: I wouldn't change that, if thats okay with you.

File: src/ProductView/TabBar/QueryInterface/Modal/Modal.js (hunk 2 above)
Comment (friedow) [reply src/ProductView/TabBar/QueryInterface/Modal/Modal.js]: Done that.

File: src/ProductView/TabBar/QueryInterface/Modal/Modal.js (hunk 3 above)
Comment (friedow) [reply src/ProductView/TabBar/QueryInterface/Modal/Modal.js]: Done that too.

File: src/RemoteHandler/DataTransmitter.js (hunk 12 above)
Comment (friedow) [reply src/RemoteHandler/DataTransmitter.js]: Done.

Comment (friedow) [reply src/RemoteHandler/DataTransmitter.js]: Done

File: src/index.js (hunk 14 above)
Comment (friedow) [reply src/index.js]: Done

**Comment by julianweise** [comment]
Please ensure that name and event query is not empty before requesting backend.

**Pull request merged**